### ISSUE TITLE
OBL_UNSATISFIED_OBLIGATION false positive

### ISSUE THREAD

**Issue by boris-petrov** [spotbugs/spotbugs#293]
The following code gives a `OBL_UNSATISFIED_OBLIGATION` warning:

```java
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;

import javax.sql.DataSource;

public class Sql {
	private final DataSource dataSource = null;

	public void executeStatement(String statementString, List<?> arguments) throws SQLException {
		try (Connection connection = dataSource.getConnection()) {
			try (PreparedStatement statement = connection.prepareStatement(statementString)) {
				statement.setFetchSize(50);
				for (int i = 0; i < arguments.size(); i++) {
					statement.setObject(i + 1, arguments.get(i));
				}

				try (ResultSet resultSet = statement.executeQuery()) {
					System.out.println(resultSet);
				}
			}
		}
	}
}
```

I think this can be regarded as a false-positive.

**Comment by aaron-haskins** [comment]
Another example of a false **OBL_UNSATISFIED_OBLIGATION** bug:

    51  public void execute(final String query) {
    52    try (Connection conn = pool.getDatabaseConnection();
    53            Statement stmt = conn.createStatement();) {
    54      stmt.execute(query);
    55    } catch (SQLException ex) {
    56      logSQLException(ex);
    57    }
    58  }

generates:

    [INFO] com.mycompany.db.Database.execute(String) may fail to clean up java.sql.Statement [com.mycompany.db.Database, com.mycompany.db.Database, com.mycompany.db.Database, com.mycompany.db.Database, com.mycompany.db.Database, com.mycompany.db.Database] Obligation to clean up resource created at Database.java:[line 53] is not dischargedPath continues at Database.java:[line 52]Path continues at Database.java:[line 54]Path continues at Database.java:[line 55]Path continues at Database.java:[line 57]Path continues at Database.java:[line 58] OBL_UNSATISFIED_OBLIGATION

**Comment by ghenkes** [comment]
I'm seeing this after upgrading to compile with Java 9 for try-with-resources statements which were (correctly) not flagged under Java 8. Maybe the bytecode try-with-resources produces has changed?

**Comment by iloveeclipse** [comment]
@ghenkes : please do not hijack other's bug reports :-)

If you observe that SpotBugs stopped to report some bug patterns on code compiled with Java 9, please open a dedicated bug with steps to reproduce.

**Comment by ghenkes** [comment]
@iloveeclipse created #432

**Comment by andrewparmet** [comment]
I'm also seeing this on Java 8 with code like: 
```java
public static MyObject forId(long myObjectId, Connection cxn) throws SQLException {
    try (PreparedStatement stmt = cxn.prepareStatement(selectMyObjectById))) {
        stmt.setLong(1, myObjectId);
        try (ResultSet res = stmt.executeQuery()) {
            res.next();
            return MyObject.fromResultSet(res);
        }
    }
}
```
where the JDBC Connection object is passed (but created with try-with-resources itself).

Is there any hope for this? I hate to scatter annotations to ignore this but I'd like to use the resource leak detector.

**Comment by vovkss** [comment]
Probably related to: https://github.com/spotbugs/spotbugs/issues/493
